src/main/java/com/glencoesoftware/omero/ms/image/region/ShapeMaskRequestHandler.java

```diff
@@ -291,42 +291,218 @@ public ShapeMaskRequestHandler(ShapeMaskCtx shapeMaskCtx, String labelImagePath,
+        query.setSubarray(getSubarray(ctx, domain));
         query.setBuffer("a1", bufferSlice);
         query.submit();
-        byte[] outputBytes = new byte[buffer.capacity()/bytesPerPixel];
+        byte[] outputBytes = new byte[capacity];
+        buffer.get(outputBytes);
+        query.close();
+        return outputBytes;
+    }
+
+    private void populateByteMask(byte[] dst, ByteBuffer src, Datatype type) {
         int i = 0;
-        while (buffer.hasRemaining()) {
-            int val = buffer.getInt(); //TODO: Handle other types
-            if (val != 0) {
-                outputBytes[i] = 1;
-            }
-            i++;
+        switch (type) {
+            case TILEDB_UINT8:
+            case TILEDB_INT8:
+                src.get(dst);
+                break;
+            case TILEDB_UINT16:
+            case TILEDB_INT16:
+                while (src.hasRemaining()) {
+                    short val = src.getShort();
+                    if (val != 0) {dst[i] = 1;}
+                    i++;
+                }
+                break;
+            case TILEDB_UINT32:
+            case TILEDB_INT32:
+                while (src.hasRemaining()) {
+                    int val = src.getInt();
+                    if (val != 0) {dst[i] = 1;}
+                    i++;
+                }
+                break;
+            case TILEDB_UINT64:
+            case TILEDB_INT64:
+                while (src.hasRemaining()) {
+                    long val = src.getLong();
+                    if (val != 0) {dst[i] = 1;}
+                    i++;
+                }
+                break;
+            default:
+                return;
         }
+    }
+
+    private byte[] getByteMaskData(Array array, Context ctx) throws TileDBError {
+        ArraySchema schema = array.getSchema();
+        Domain domain = schema.getDomain();
+        Attribute attribute = schema.getAttribute("a1");
+
+        int bytesPerPixel = 1;
+        switch (attribute.getType()) {
+            case TILEDB_UINT8:
+            case TILEDB_INT8:
+                break;
+            case TILEDB_UINT16:
+            case TILEDB_INT16:
+                bytesPerPixel = 2;
+                break;
+            case TILEDB_UINT32:
+            case TILEDB_INT32:
+                bytesPerPixel = 4;
+                break;
+            case TILEDB_UINT64:
+            case TILEDB_INT64:
+                bytesPerPixel = 8;
+                break;
+            default:
+                throw new IllegalArgumentException("Attribute type " + attribute.getType().toString() + " not supported");
+        }
+        int capacity = getCapacity(domain) * bytesPerPixel;
+        ByteBuffer buffer = ByteBuffer.allocateDirect(capacity);
+        ByteBuffer bufferSlice = buffer.slice();
+        bufferSlice.order(ByteOrder.nativeOrder());
+        //Dimensions in Dense Arrays must be the same type
+        Query query = new Query(array, QueryType.TILEDB_READ);
```

Comment:
As above. You are allocating native memory here and this should always be in a try with resource block.